extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/apache/guacamole/auth/jdbc/tunnel/AccessEnforcingDelegatingTunnel.java

```diff
@@ -0,0 +1,196 @@
+import org.apache.guacamole.auth.jdbc.user.UserService;
+import org.apache.guacamole.io.GuacamoleReader;
+import org.apache.guacamole.net.DelegatingGuacamoleTunnel;
+import org.apache.guacamole.net.GuacamoleTunnel;
+import org.apache.guacamole.protocol.FilteredGuacamoleReader;
+import org.apache.guacamole.protocol.GuacamoleFilter;
+import org.apache.guacamole.protocol.GuacamoleInstruction;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.inject.Inject;
+import com.google.inject.assistedinject.Assisted;
+import com.google.inject.assistedinject.AssistedInject;
+
+/**
+ * A tunnel implementation that enforces access window restriction for the
+ * provided ModeledUser, throwing a GuacamoleUnauthorizedException if the
+ * user's configured access window has closed, or if the user has become
+ * disabled. All other tunnel implementation is delegated to the underlying
+ * tunnel object.
+ */
+public class AccessEnforcingDelegatingTunnel extends DelegatingGuacamoleTunnel {
+
+    /**
+     * Logger for this class.
+     */
+    private static final Logger logger = LoggerFactory.getLogger(
+            AccessEnforcingDelegatingTunnel.class);
+
+    /**
+     * The number of milliseconds between subsequent refreshes of the user
+     * from the DB.
+     */
+    private static final long USER_MODEL_REFRESH_INTERVAL = 10000;
+
+    /**
+     * The user who's access window restrictions should be applied for the
+     * wrapped tunnel.
+     */
+    private final AtomicReference<ModeledUser> user;
+
+    /**
+     * A thread that will continously refresh the user
+     */
+    private final Thread userRefreshThread;
+
+    /**
+     * A service to use for refreshing the user from the DB.
+     */
+    @Inject
+    private UserService userService;
+
+    /**
+     * Create a new tunnel that will enforce the access window restrictions of
+     * the provided user, during usage of the provided tunnel.
+     *
+     * @param tunnel
+     *     The tunnel to delegate to.
+     *
+     * @param modeledAuthenticatedUser
+     *     The user whose access restrictions should be applied.
+     *
+     */
+    @AssistedInject
+    public AccessEnforcingDelegatingTunnel(
+            @Nonnull @Assisted GuacamoleTunnel tunnel,
+            @Nonnull @Assisted ModeledAuthenticatedUser modeledAuthenticatedUser) {
+
+        super(tunnel);
+        this.user = new AtomicReference<>(modeledAuthenticatedUser.getUser());
+
+        this.userRefreshThread = new Thread(() -> {
+            while (true) {
+
+                try {
+
+                    // Fetch an up-to-date user record from the DB to ensure
+                    // that any access restrictions modified while this tunnel
+                    // is open will be taken into account
+                    this.user.set(userService.retrieveUser(
+                        modeledAuthenticatedUser.getAuthenticationProvider(),
+                        modeledAuthenticatedUser));
```

Comment:
I think this might cause problems at scale if we're executing a SQL query every 10 seconds for every connection across every user. It should instead occur once across all logged-in users (ie: at the `AuthenticationProvider` level), presumably at most once per minute.